splicemachine/ml/stats.py

```diff
@@ -132,6 +240,262 @@ def _transform(self, dataset):
+
+                for (minorityLabel, ratio) in ratios.items():
+                    minorityData = dataset.filter(F.col(self.labelCol) == minorityLabel).sample(withReplacement = self.withReplacement, fraction = ratio, seed = self.randomState)
+                    if not returnData:
+                        returnData = majorityData.union(minorityData)
+                    else:
+                        returnData = returnData.union(minorityData)
+
+            return returnData
+        else:
+            raise NotImplementedError("Only auto is currently implemented")
+
+class OverSampleCrossValidator(CrossValidator):
+    """Class to perform Cross Validation model evaluation while over-sampling minority labels.
+
+    Example:
+    -------
+    >>> from pyspark.sql.session import SparkSession
+    >>> from pyspark.ml.classification import LogisticRegression
+    >>> from pyspark.ml.evaluation import BinaryClassificationEvaluator, MulticlassClassificationEvaluator
+    >>> from pyspark.ml.linalg import Vectors
+    >>> from splicemachine.ml.stats import OverSampleCrossValidator
+    >>> spark = SparkSession.builder.getOrCreate()
+    >>> dataset = spark.createDataFrame(
+    ...      [(Vectors.dense([0.0]), 0.0),
+    ...       (Vectors.dense([0.5]), 0.0),
+    ...       (Vectors.dense([0.4]), 1.0),
+    ...       (Vectors.dense([0.6]), 1.0),
+    ...       (Vectors.dense([1.0]), 1.0)] * 10,
+    ...      ["features", "label"])
+    >>> lr = LogisticRegression()
+    >>> grid = ParamGridBuilder().addGrid(lr.maxIter, [0, 1]).build()
+    >>> PRevaluator = BinaryClassificationEvaluator(metricName = 'areaUnderPR')
+    >>> AUCevaluator = BinaryClassificationEvaluator(metricName = 'areaUnderROC')
+    >>> ACCevaluator = MulticlassClassificationEvaluator(metricName="accuracy")
+    >>> cv = OverSampleCrossValidator(estimator=lr, estimatorParamMaps=grid, evaluator=AUCevaluator, altEvaluators = [PRevaluator, ACCevaluator],parallelism=2,seed = 1234)
+    >>> cvModel = cv.fit(dataset)
+    >>> print(cvModel.avgMetrics)
+    [(0.5, [0.5888888888888888, 0.3888888888888889]), (0.806878306878307, [0.8556863149300125, 0.7055555555555556])]
+    >>> print(AUCevaluator.evaluate(cvModel.transform(dataset)))
+    0.8333333333333333
+    """
+    def __init__(self, estimator=None, estimatorParamMaps=None, evaluator=None, numFolds=3, seed=None, parallelism=1, collectSubModels=False, labelCol = 'label', altEvaluators = None, overSample = True):
+        """ Initialize Self
+
+        :param estimator: Machine Learning Model, defaults to None
+        :param estimatorParamMaps: paramMap to search, defaults to None
+        :param evaluator: primary model evaluation metric, defaults to None
+        :param numFolds: number of folds to perform, defaults to 3
+        :param seed: random state, defaults to None
+        :param parallelism: number of threads, defaults to 1
```

Comment:
Should parallelism default to the same as numFolds (3)?

Reply:
That is a good point